sdks/java/io/google-ads/src/main/java/org/apache/beam/sdk/io/googleads/GoogleAdsV14.java

```diff
@@ -0,0 +1,647 @@
+ * from a query using {@link Read#withQuery(String)} and one or a few customer IDs using either
+ * {@link Read#withCustomerId(Long)} or {@link Read#withCustomerIds(List)}. Alternatively, use
+ * {@link GoogleAdsV14#readAll()} to read either a bounded or unbounded {@link PCollection} of
+ * {@link GoogleAdsRow} from a {@link PCollection} of {@link SearchGoogleAdsStreamRequest}.
+ *
+ * <p>For example, using {@link GoogleAdsV14#read()}:
+ *
+ * <pre>{@code
+ * Pipeline p = Pipeline.create();
+ * PCollection<GoogleAdsRow> rows =
+ *     p.apply(
+ *         GoogleAdsIO.v14()
+ *             .read()
+ *             .withCustomerId(1234567890l)
+ *             .withQuery(
+ *                 "SELECT"
+ *                     + "campaign.id,"
+ *                     + "campaign.name,"
+ *                     + "campaign.status"
+ *                     + "FROM campaign"));
+ * p.run();
+ * }</pre>
+ *
+ * <p>Alternatively, using {@link GoogleAdsV14#readAll()} to execute requests from a {@link
+ * PCollection} of {@link SearchGoogleAdsStreamRequest}:
+ *
+ * <pre>{@code
+ * Pipeline p = Pipeline.create();
+ * PCollection<SearchGoogleAdsStreamRequest> requests =
+ *     p.apply(
+ *         Create.of(
+ *             ImmutableList.of(
+ *                 SearchGoogleAdsStreamRequest.newBuilder()
+ *                     .setCustomerId(Long.toString(1234567890l))
+ *                     .setQuery(
+ *                         "SELECT"
+ *                             + "campaign.id,"
+ *                             + "campaign.name,"
+ *                             + "campaign.status"
+ *                             + "FROM campaign")
+ *                     .build())));
+ * PCollection<GoogleAdsRow> rows = requests.apply(GoogleAdsIO.v14().readAll());
+ * p.run();
+ * }</pre>
+ *
+ * <h2>Client-side rate limiting</h2>
+ *
+ * On construction of a {@link GoogleAdsV14#read()} or {@link GoogleAdsV14#readAll()} transform a
+ * default rate limiting policy is provided to stay well under the rate limit for the Google Ads
+ * API, but this limit is only local to a single worker and operates without any knowledge of other
+ * applications using the same developer token for any customer ID. The Google Ads API enforces
+ * global limits from the developer token down to the customer ID and it is recommended to host a
+ * shared rate limiting service to coordinate traffic to the Google Ads API across all applications
+ * using the same developer token. Users of these transforms are strongly advised to implement their
+ * own {@link RateLimitPolicy} and {@link RateLimitPolicyFactory} to interact with a shared rate
+ * limiting service for any production workloads.
+ *
+ * @see GoogleAdsIO#v14()
+ * @see GoogleAdsOptions
+ * @see <a href="https://developers.google.com/google-ads/api/docs/best-practices/overview">Best
+ *     Practices in the Google Ads documentation</a>
+ */
+public class GoogleAdsV14 {
```

Comment:
By convention, we wouldn't split out our IOs by API version, and if we wanted to update to some new API version, we would do that in a way that was ideally invisible to the users.

I'm not familiar enough with the Google Ads API to know if that is reasonable or not, but I thought I would bring it up here.

Reply:
The Google Ads API updates, deprecates, and sunsets its API versions (major versions have a 12 months support cycle) quite a bit more frequently than the Cloud Platform APIs do, see https://developers.google.com/google-ads/api/docs/sunset-dates. The output of the IO's Read/ReadAll transforms is a fairly complex protobuf which can describe every possible reporting row. Similarly, the planned write transforms will use e.g. https://developers.google.com/google-ads/api/reference/rpc/v14/GoogleAdsService#mutate to mutate objects, taking as input a protobuf which can describe nearly every possible mutation.

Upgrades between API versions are not trivial either, some Google Ads features may be remapped to different services or removed entirely, see https://developers.google.com/google-ads/api/docs/upgrade. Unfortunately, this makes refactoring to a new API version more of a hassle for the user than simply replacing the version namespace everywhere.

My thought on splitting the IO by version was that it communicates to the user and readers which API version's concepts and quirks they can expect without having to look at the transform's input or output type. Users may also need to compare results from new and old API versions if an API concept has changed radically between versions.

Reply:
That makes sense to me. It isn't ideal within the beam model, but it makes sense for this case